python/hail/expr/functions.py

```diff
@@ -3114,6 +3172,41 @@ def sum(collection, filter_missing: bool = True) -> NumericExpression:
     return collection._filter_missing_method(filter_missing, "sum", collection.dtype.element_type)
 
 
+@typecheck(a=expr_array(expr_numeric),
+            filter_missing=bool)
+def cumulative_sum(a, filter_missing: bool = True) -> ArrayNumericExpression:
+    """Returns an array of the cumulative sum of values in the array.
+
+    Examples
+    --------
+
+    >>> a = [1, 3, 5, 6, 7, 9]
+
+    >>> hl.cumulative_sum(a).value
+    [1, 4, 9, 15, 22, 31]
+
+    Note
+    ----
+    Missing elements are ignored if `filter_missing` is ``True``. If `filter_missing`
+    is ``False``, then any missing element causes the result to be missing.
+
+    Parameters
+    ----------
+    a : :class:`.ArrayNumericExpression`
+        Array expression with numeric element type.
+    filter_missing : :obj:`bool`
+        Remove missing elements from the collection before computing product.
+
+    Returns
+    -------
+    :class:`.ArrayNumericExpression`
+    """
+    if filter_missing:
+        a = a.filter(hl.is_defined)
+    return a.scan(hl.literal(0, a.dtype.element_type),
```

Comment:
I think that seems reasonable, as long as evaluating f twice is relatively fast/painless

Reply:
it's just in Python - I'm not worried about time so much as weird errors getting generated.

Reply:
though we can check it as follows: 

```
evaluate f

if body.dtype == zero.dtype:
  proceed
else:
  coercer = coercer.from_dtype(body.dtype) # or whatever this is
  if coercer.can_coerce(zero.dtype):  # or whatever
    coerced_zero = coercer.coerce(zero)
    re evaluate body with coerced_zero and proceed
  else:
    raise the types-don't-match error
```

Reply:
ah, here's the potential problem -- what if evaluating body with `coerced_zero` produces a different type? I guess if it does that we can panic to the types-dont-match error

Reply:
yep. I now have it trying to coerce `body` to the zero type first, since that doesn't require re-evaluation of `f`, and then trying to coerce `zero`, and checking if the new `body` is coercable to the new `zero`. Hopefully that catches common conversion problems without getting stuck in a loop trying to coerce `zero` and re-evaluate `f`, but if it doesn't work I throw an error with the original types.